packages/frontend/app/services/channels/channel-verifier.ts

```diff
@@ -0,0 +1,130 @@
+import Service, { inject as service } from '@ember/service';
+import ChannelContextChain from 'emberclear/models/channel-context-chain';
+import VoteChain, { VOTE_ACTION } from 'emberclear/models/vote-chain';
+import VoteVerifier from 'emberclear/services/channels/vote-verifier';
+import Identity from 'emberclear/models/identity';
+import { identityEquals, identitiesIncludes } from 'emberclear/utils/identity-comparison';
+
+export default class ChannelVerifier extends Service {
+  @service('channels/vote-verifier') voteVerifier!: VoteVerifier;
+
+  async isValidChain(channel: ChannelContextChain): Promise<boolean> {
+    if (channel.previousChain === undefined) {
+      return this.isValidSingleChain(channel);
+    }
+
+    let somethingWrongWithPastOrSupportingVote =
+      !(await this.isValidChain(channel.previousChain)) ||
+      !(await this.voteVerifier.verify(channel.supportingVote)) ||
+      !this.isVoteCompletedPositive(channel.supportingVote, channel.previousChain.admin);
+
+    if (somethingWrongWithPastOrSupportingVote) {
+      return false;
+    }
+
+    switch (channel.supportingVote.action) {
+      case VOTE_ACTION.ADD:
+        return this.isAddValid(channel);
+      case VOTE_ACTION.PROMOTE:
+        return this.isPromoteValid(channel);
+      case VOTE_ACTION.REMOVE:
+        return this.isRemoveValid(channel);
+      default:
+        return false;
+    }
+  }
+
+  private isValidSingleChain(channel: ChannelContextChain): boolean {
+    return !(
+      channel.admin === undefined ||
+      channel.members.length !== 1 ||
+      !identityEquals(channel.members.objectAt(0)!, channel.admin)
+    );
+  }
+
+  private isVoteCompletedPositive(vote: VoteChain, admin: Identity): boolean {
+    if (
+      vote.yes.length > vote.no.length + vote.remaining.length ||
+      (vote.yes.length === vote.no.length + vote.remaining.length &&
+        identitiesIncludes(vote.yes.toArray(), admin))
+    ) {
+      return true;
+    }
+    return false;
+  }
```

Comment:
i feel like this is going to be used in more places. are we planning to have a democracy utility library sort of thing somewhere?

Reply:
or is this planned to be it

Reply:
This method in particular, once needed somewhere else, will be moved into the channels/utils package, made a public export as it needs no state, and unit tested separately. For now though it is staying here.